command/agent/fs_endpoint.go

```diff
@@ -70,19 +71,44 @@ func (s *HTTPServer) FsRequest(resp http.ResponseWriter, req *http.Request) (int
 		return nil, clientNotRunning
 	}
 
+	var secret string
+	s.parseToken(req, &secret)
+
+	var namespace string
+	parseNamespace(req, &namespace)
+
+	requireReadFS := func(f func(http.ResponseWriter, *http.Request) (interface{}, error)) (interface{}, error) {
+		if aclObj, err := s.agent.Client().ResolveToken(secret); err != nil {
+			return nil, err
+		} else if aclObj != nil && !aclObj.AllowNsOp(namespace, acl.NamespaceCapabilityReadFS) {
+			return nil, structs.ErrPermissionDenied
+		}
+		return f(resp, req)
+	}
+
 	path := strings.TrimPrefix(req.URL.Path, "/v1/client/fs/")
 	switch {
 	case strings.HasPrefix(path, "ls/"):
-		return s.DirectoryListRequest(resp, req)
+		return requireReadFS(s.DirectoryListRequest)
 	case strings.HasPrefix(path, "stat/"):
-		return s.FileStatRequest(resp, req)
+		return requireReadFS(s.FileStatRequest)
 	case strings.HasPrefix(path, "readat/"):
-		return s.FileReadAtRequest(resp, req)
+		return requireReadFS(s.FileReadAtRequest)
 	case strings.HasPrefix(path, "cat/"):
-		return s.FileCatRequest(resp, req)
+		return requireReadFS(s.FileCatRequest)
 	case strings.HasPrefix(path, "stream/"):
-		return s.Stream(resp, req)
+		return requireReadFS(s.Stream)
 	case strings.HasPrefix(path, "logs/"):
+		// Logs can be accessed with ReadFS or ReadLogs caps
+		if aclObj, err := s.agent.Client().ResolveToken(secret); err != nil {
```

Comment:
Can you just resolve the token and check the error at line 79 and then the helper func and this just do the else if part.